www/js/controller.js

```diff
@@ -692,6 +700,335 @@ define([
+					this.updateInputNode(node.id, node.variable);
+				}));
+
+				var variableList = parse.lhsParse.variables().concat(parse.rhsParse.variables());
+				array.forEach(variableList, lang.hitch(this, function(n){
+					this.nodeConnections.push(n);
+				}));
+
+				// Send to PM if all variables are known.
+				console.log(parse);
+			}
+		},
+
+		substituteVars : function(parse,ignoreUnknownTest){
+		//getDescriptionID is present only in student mode. So in author mode it will give an identity function. This is a work around in case when its in author mode at that time the given model is the actual model. So descriptionID etc are not available. 
+			var mapID = this._model.active.getAuthoredID || function(x){ return x; };
+			var unMapID = this._model.active.getNodeIDFor || function(x){ return x; };
+			var currentNodelist = [];
+			array.forEach(parse.variables(), function(variable){
+					// Test if variable name can be found in given model
+					console.log("each variable",variable);
+					var givenID = this._model.authored.getNodeIDByName(variable);
+
+					// autocreation flag will eventually be set from the URL
+					// or from the state table.  Alternatively, we will show a list
+					// of variables.
+					var autocreationFlag = true;
+					/*if(autocreationFlag && !this._model.active.isNode(givenID))
+					 this.autocreateNodes(variable);*/
+					// The variable "descriptionID" is the corresponding givenModelNodeID from the model (it is not equal to the givenID used here).
+					// The variable "badVarCount" is used to track the number of times a user has attempted to use an incorrect variable to prevent
+					//		him or her from being stuck indefinitely.
+
+					var descriptionID = "";
+					//var badVarCount = "";
+					/*if (!ignoreUnknownTest) {
+						// Check for number of unknown var, only in student mode.
+						descriptionID = this._model.active.getDescriptionID(this.currentID);
+						badVarCount = this._model.given.getAttemptCount(descriptionID, "unknownVar");
+					}*/
+
+					if(givenID || ignoreUnknownTest){
+						//|| ignoreUnknownTest || badVarCount > 3){
+						// Test if variable has been defined already
+						var subID = unMapID.call(this._model.active, givenID);
+						if(subID){
+							// console.log("	   substituting ", variable, " -> ", studentID);
+							parse.substitute(variable, subID);
+							//inputNodesList.push({ "id": subID, "variable":variable});
+						}else if(autocreationFlag){
+							//create node
+							var id = this._model.active.addNode();
+							this.addNode(this._model.active.getNode(id));
+							this.setNodeDescription(id, variable);
+
+							currentNodelist.push({ "id": id, "variable":variable});
+							//get Node ID and substitute in equation
+							var subID2 = unMapID.call(this._model.active, givenID||id);
+							parse.substitute(variable, subID2); //this should handle createInputs and connections to automatic node
+						}/*else{
```

Comment:
Yeah substitute is the funciton, but would it be possible to remodularize this so that auto create nodes work and there is a function in equation.js which can be called to replace names with their corresponding ids. It takes an input of the complete equation and replaces them with their ids. So basically you will parse it auto create the new nodes, and then send the equation to that function in equation.js which will return the equation with ids.